lxd/acme/acme.go

```diff
@@ -0,0 +1,191 @@
+)
+
+// retries describes the number of retries after which LXD will give up registering a user or
+// issuing a new certificate. The number 5 was chosen because Let's Encrypt has a limit of 5
+// failures per account, per hostname, per hour.
+const retries = 5
+
+// ClusterCertFilename describes the filename of the new certificate which is stored in case it
+// cannot be distributed in a cluster due to offline members. LXD will try to distribute this
+// certificate at a later stage.
+const ClusterCertFilename = "cluster.crt.new"
+
+func certificateNeedsUpdate(domain string, cert *x509.Certificate) bool {
+	return shared.StringInSlice(domain, cert.DNSNames) && !time.Now().After(cert.NotAfter.Add(-time.Hour*24*30))
+}
+
+// UpdateCertificate updates the certificate.
+func UpdateCertificate(s *state.State, provider HTTP01Provider, clustered bool, domain string, email string, caURL string) (*certificate.Resource, error) {
+	clusterCertFilename := shared.VarPath(ClusterCertFilename)
+
+	l := logger.AddContext(logger.Log, logger.Ctx{"domain": domain, "caURL": caURL})
+
+	// If clusterCertFilename exists, it means that a previously issued certificate couldn't be
+	// distributed to all cluster members and was therefore kept back. In this case, don't issue
+	// a new certificate but return the previously issued one.
+	if clustered && shared.PathExists(clusterCertFilename) {
+		keyFilename := shared.VarPath("cluster.key")
+
+		clusterCert, err := os.ReadFile(clusterCertFilename)
+		if err != nil {
+			return nil, fmt.Errorf("Failed reading cluster certificate file: %w", err)
+		}
+
+		key, err := os.ReadFile(keyFilename)
+		if err != nil {
+			return nil, fmt.Errorf("Failed reading cluster key file: %w", err)
+		}
+
+		keyPair, err := tls.X509KeyPair(clusterCert, key)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get keypair: %w", err)
+		}
+
+		cert, err := x509.ParseCertificate(keyPair.Certificate[0])
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse certificate: %w", err)
+		}
+
+		// If the domain matches the certificate's DNS names, and the certificate is valid for more
+		// than 30 days, return it.
+		if certificateNeedsUpdate(domain, cert) {
+			return &certificate.Resource{
+				Certificate: clusterCert,
+				PrivateKey:  key,
+			}, nil
+		}
+	}
+
+	// Load the certificate.
+	certInfo, err := util.LoadCert(s.OS.VarDir)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load certificate and key file: %w", err)
+	}
+
+	cert, err := x509.ParseCertificate(certInfo.KeyPair().Certificate[0])
+	if err != nil {
+		return nil, fmt.Errorf("Failed to parse certificate: %w", err)
+	}
+
+	// Check the certificate's expiration date, and renew it if it's only valid for less than 30 days.
+	// This is only done if our certificate's DNS names contains the request domain.
+	if certificateNeedsUpdate(domain, cert) {
+		l.Debug("Skipping certificate renewal as it is still valid for more than 30 days")
```

Comment:
Maybe `certificateNeedsUpdate` needs to return an error instead which can be used here, as the reason it may need an update may be because of a mismatch of domains, not just the expiry date.

Reply:
Actually, the function logic is wrong. It should return `true` on domain mismatch OR "expired" certificate.